src/github.com/matrix-org/go-neb/services/feedreader/feedreader.go

```diff
@@ -0,0 +1,235 @@
+func (p *feedPoller) queryFeed(s *feedReaderService, feedURL string) (*gofeed.Feed, []gofeed.Item, error) {
+	log.WithField("feed_url", feedURL).Info("Querying feed")
+	var items []gofeed.Item
+	fp := gofeed.NewParser()
+	feed, err := fp.ParseURL(feedURL)
+	if err != nil {
+		return nil, items, err
+	}
+
+	// Work out which items are new, if any (based on the last updated TS we have)
+	// If the TS is 0 then this is the first ever poll, so let's not send 10s of events
+	// into the room and just do new ones from this point onwards.
+	if s.Feeds[feedURL].FeedUpdatedTimestampSecs != 0 {
+		for _, i := range feed.Items {
+			if i == nil || i.PublishedParsed == nil {
+				continue
+			}
+			if i.PublishedParsed.Unix() > s.Feeds[feedURL].FeedUpdatedTimestampSecs {
+				items = append(items, *i)
+			}
+		}
+	}
+
+	now := time.Now().Unix() // Second resolution
+
+	// Work out when this feed was last updated
+	var feedLastUpdatedTs int64
+	if feed.UpdatedParsed != nil {
+		feedLastUpdatedTs = feed.UpdatedParsed.Unix()
+	} else if len(feed.Items) > 0 {
+		i := feed.Items[0]
+		if i != nil && i.PublishedParsed != nil {
+			feedLastUpdatedTs = i.PublishedParsed.Unix()
+		}
+	}
+
+	// Work out when to next poll this feed
+	nextPollTsSec := now + minPollingIntervalSeconds
+	if s.Feeds[feedURL].PollIntervalMins > int(minPollingIntervalSeconds/60) {
+		nextPollTsSec = now + int64(s.Feeds[feedURL].PollIntervalMins*60)
+	}
+	// TODO: Handle the 'sy' Syndication extension to control update interval.
+	// See http://www.feedforall.com/syndication.htm and http://web.resource.org/rss/1.0/modules/syndication/
+
+	p.updateFeedInfo(s, feedURL, nextPollTsSec, feedLastUpdatedTs)
+	return feed, items, nil
+}
+
+func (p *feedPoller) updateFeedInfo(s *feedReaderService, feedURL string, nextPollTs, feedUpdatedTs int64) {
+	for u := range s.Feeds {
+		if u != feedURL {
+			continue
+		}
+		f := s.Feeds[u]
+		f.NextPollTimestampSecs = nextPollTs
+		f.FeedUpdatedTimestampSecs = feedUpdatedTs
+		s.Feeds[u] = f
+	}
+}
+
+func (p *feedPoller) sendToRooms(s *feedReaderService, cli *matrix.Client, feedURL string, feed *gofeed.Feed, item gofeed.Item) error {
+	logger := log.WithField("feed_url", feedURL).WithField("title", item.Title)
+	logger.Info("New feed item")
+	var rooms []string
+	for roomID, urls := range s.Rooms {
+		for _, u := range urls {
+			if u == feedURL {
+				rooms = append(rooms, roomID)
+				break
+			}
+		}
+	}
+	for _, roomID := range rooms {
+		if _, err := cli.SendMessageEvent(roomID, "m.room.message", itemToHTML(feed, item)); err != nil {
+			logger.WithError(err).WithField("room_id", roomID).Error("Failed to send to room")
+		}
+	}
+	return nil
+}
+
+// SomeOne posted a new article: Title Of The Entry ( https://someurl.com/blag )
+func itemToHTML(feed *gofeed.Feed, item gofeed.Item) matrix.HTMLMessage {
+	return matrix.GetHTMLMessage("m.notice", fmt.Sprintf(
+		"<i>%s</i> posted a new article: %s ( %s )",
```

Comment:
We don't currently support templates in Go-NEB. I think we will want to in the future, so people can configure their GH messages, RSS messages, JIRA messages, etc. But that doesn't exist yet.